core-js/src/definitions.ts

```diff
@@ -0,0 +1,1459 @@
+  addDirectiveDefinition(directive: DirectiveDefinition): DirectiveDefinition; addDirectiveDefinition(directiveOrName: string | DirectiveDefinition): DirectiveDefinition {
+    const definition = typeof directiveOrName === 'string' ? new DirectiveDefinition(directiveOrName) : directiveOrName;
+    if (this.directive(definition.name)) {
+      throw buildError(`Directive ${definition} already exists in this schema`);
+    }
+    if (definition.parent) {
+      // For convenience, let's not error out on adding an already added directive.
+      if (definition.parent == this) {
+        return definition;
+      }
+      throw buildError(`Cannot add directive ${definition} to this schema; it is already attached to another schema`);
+    }
+    if (definition.isBuiltIn) {
+      if (!this.isConstructed) {
+        this._builtInDirectives.set(definition.name, definition);
+      } else {
+        throw buildError(`Cannot add built-in ${definition} to this schema (built-ins can only be added at schema construction time)`);
+      }
+    } else {
+      this._directives.set(definition.name, definition);
+    }
+    SchemaElement.prototype['setParent'].call(definition, this);
+    return definition;
+  }
+
+  clone(builtIns?: BuiltIns): Schema {
+    const cloned = new Schema(builtIns ?? this.builtIns);
+    copy(this, cloned);
+    return cloned;
+  }
+}
+
+export class SchemaDefinition extends SchemaElement<Schema, never>  {
+  readonly kind: 'SchemaDefinition' = 'SchemaDefinition';
+  protected readonly _roots: Map<SchemaRoot, ObjectType> = new Map();
+
+  get coordinate(): string {
+    return '';
+  }
+
+  get roots(): ReadonlyMap<SchemaRoot, ObjectType> {
+    return this._roots;
+  }
+
+  root(rootType: SchemaRoot): ObjectType | undefined {
+    return this._roots.get(rootType);
+  }
+
+  setRoot(rootType: SchemaRoot, type: ObjectType): ObjectType;
+  setRoot(rootType: SchemaRoot, name: string, source?: ASTNode): ObjectType;
+  setRoot(rootType: SchemaRoot, nameOrType: ObjectType | string, source?: ASTNode): ObjectType {
+    let toSet: ObjectType;
+    if (typeof nameOrType === 'string') {
+      this.checkUpdate();
+      const obj = this.schema()!.type(nameOrType);
+      if (!obj) {
+        throw new GraphQLError(`Cannot set schema ${rootType} root to unknown type ${nameOrType}`, source);
+      } else if (obj.kind != 'ObjectType') {
+        throw new GraphQLError(`Cannot set schema ${rootType} root to non-object type ${nameOrType} (of type ${obj.kind})`, source);
+      }
+      toSet = obj;
+    } else {
+      this.checkUpdate(nameOrType);
+      toSet = nameOrType;
+    }
+    this._roots.set(rootType, toSet);
+    this.source = source;
+    addReferenceToType(this, toSet);
+    return toSet;
+  }
+
+  protected removeTypeReference(toRemove: NamedType) {
+    for (const [root, type] of this._roots) {
+      if (type == toRemove) {
+        this._roots.delete(root);
+      }
+    }
+  }
+
+  remove(): never[] {
```

Comment:
I don't think there's a need to support removing the schema definition. That also avoids having to implement/invoke `resetSchemaDefinition`.

Reply:
This was mostly done due to `remove` being abstract in `SchemaElement` and `SchemaDefinition` being a `SchemaElement`. With the change to `NamedSchemaElement`, I moved the `remove` method there and thus removed `SchemaDefinition.remove()`.